0025-nat-traversal.adoc

```diff
@@ -0,0 +1,174 @@
+
+NAT: Network address translation/translator
+
+UPnP: Universal plug and play
+
+TURN: Traveral using relays around NAT
+
+STUN: Session traversal utilities around NAT
+
+== Context
+
+NAT allows the use of private IP addresses on private networks behind routers with a single public IP address facing the Internet.
+This can make it difficult for a comit node to determine a public address that can be used to receive inbound connections.
+This documents evaluates a number of NAT traversal methods that allow a comit node to connect to another comit node situated behind a NAT.
+
+== Goals
+
+1. Allow a comit node to traverse a NAT to reach another comit node.
+
+== NAT Traversal Methods
+
+This sections provides an overview and evaluation of current NAT traversal methods.
+In these examples, unless stated otherwise, Alice is dialing Bob who is behind a NAT.
+
+=== Manual Port Forwarding
+
+Bob configures his NAT to forward data received from an publicly facing port and IP address to the private IP address of device.
+Bob notifies Alice that he can reached through the aforementioned publicly facing port and IP address.
+
+This solution does not require any changes to the COMIT codebase and places the burden on the user to configure their NAT.
+Users who do not have control over the configuration of their NAT can not use this method.
+We cannot completely rely on port forwarding because users on mobile, shared or corporate managed networks will not have access to NAT configuration.
+
+- + Alice and Bob can communicate through a direct connection
+
+- - Requires Bob to have access to NAT configuration
+
+- - Requires Bob to configure his NAT
+
+
+=== Automatic Port Forwarding using UPnP
+
+A UPnP enabled NAT automatically forwards publicly facing ports as requested by an application.
+Unlike manual port forwarding, Bob is not required to configure his NAT.
+Not all NATs are UPnP capable and it is often disabled or not used by peer to peer applications.
+Bitcoin core removed UPnP support due to security concerns: https://bitcoin.org/en/alert/2015-10-12-upnp-vulnerability.
+
+- + Does not require Bob to configure his NAT
+
+- + Alice and Bob can communicate through a direct connection
+
+- - Requires the user to have a UPnP capable and enabled NAT (UPnP capable routers are fairly common in home networks)
+
+- - Opens up the user to security vulnerabilities arising from the NAT blindly trusting applications and allowing to open externally facing ports
+
+=== Reverse connection
+
+Instead of Alice attempting to reach Bob through his NAT. Bob dials Alice and Alice uses the connection to send to data to Bob.
+This technique only works if Alice knows her public IP and port and has provided it to Bo.
+Since Alice is unable to dial Bob, Alice is unable to initiate communication and must wait to be dialled by a Bob.
+
+Reverse connections cannot be used because many of our protocols require both peers to be able to initate a connection with each other.
```

Comment:
That is not quite true.

Libp2p is specifically designed so you can open substreams in either direction without having to care about the physical connection.

If for some reason we already have a physical connection to a node, then any of the peers can use this with any kind of protocol on top.